From: You-Sheng Yang <vicamo@gmail.com>
Date: Wed, 28 May 2025 16:56:29 +0800
Subject: wl: use timer_delete for kernel >= 6.15

`del_timer{,_sync}` have been redirected to `timer_delete{,_sync}` since
v6.2-rc1 and they are now removed in v6.15-rc1 commit 8fa7292fee5c5
("treewide: Switch/rename to timer_delete[_sync]()").

Rename of del_timer_sync was backported to v6.1.83, and del_timer to
v6.1.91.

Bug-Ubuntu: https://bugs.launchpad.net/bugs/2111898
Signed-off-by: You-Sheng Yang <vicamo@gmail.com>
---
 amd64/src/include/linuxver.h | 18 ++++++++++++++++--
 amd64/src/wl/sys/wl_linux.c  |  2 +-
 i386/src/include/linuxver.h  | 18 ++++++++++++++++--
 i386/src/wl/sys/wl_linux.c   |  2 +-
 4 files changed, 34 insertions(+), 6 deletions(-)

diff --git a/amd64/src/include/linuxver.h b/amd64/src/include/linuxver.h
index 79557c3..bbe3cb9 100644
--- a/amd64/src/include/linuxver.h
+++ b/amd64/src/include/linuxver.h
@@ -346,14 +346,28 @@ static inline void tasklet_init(struct tasklet_struct *tasklet,
 }
 #define tasklet_kill(tasklet)	{ do {} while (0); }
 
-#define del_timer_sync(timer) del_timer(timer)
-
 #else
 
 #define netif_down(dev)
 
 #endif 
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2, 3, 43))
+
+#define timer_delete(timer)             del_timer(timer)
+#define timer_delete_sync(timer)        del_timer(timer)
+
+#elif (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 83))
+
+#define timer_delete(timer)             del_timer(timer)
+#define timer_delete_sync(timer)        del_timer_sync(timer)
+
+#elif (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 91))
+
+#define timer_delete(timer)             del_timer(timer)
+
+#endif
+
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 3))
 
 #define PREPARE_TQUEUE(_tq, _routine, _data)			\
diff --git a/amd64/src/wl/sys/wl_linux.c b/amd64/src/wl/sys/wl_linux.c
index 7c684c4..cbf4efe 100644
--- a/amd64/src/wl/sys/wl_linux.c
+++ b/amd64/src/wl/sys/wl_linux.c
@@ -2482,7 +2482,7 @@ wl_del_timer(wl_info_t *wl, wl_timer_t *t)
 	ASSERT(t);
 	if (t->set) {
 		t->set = FALSE;
-		if (!del_timer(&t->timer)) {
+		if (!timer_delete(&t->timer)) {
 #ifdef BCMDBG
 			WL_INFORM(("wl%d: Failed to delete timer %s\n", wl->unit, t->name));
 #endif
diff --git a/i386/src/include/linuxver.h b/i386/src/include/linuxver.h
index b05bc32..35fc1c5 100644
--- a/i386/src/include/linuxver.h
+++ b/i386/src/include/linuxver.h
@@ -341,14 +341,28 @@ static inline void tasklet_init(struct tasklet_struct *tasklet,
 }
 #define tasklet_kill(tasklet)	{ do {} while (0); }
 
-#define del_timer_sync(timer) del_timer(timer)
-
 #else
 
 #define netif_down(dev)
 
 #endif 
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2, 3, 43))
+
+#define timer_delete(timer)             del_timer(timer)
+#define timer_delete_sync(timer)        del_timer(timer)
+
+#elif (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 83))
+
+#define timer_delete(timer)             del_timer(timer)
+#define timer_delete_sync(timer)        del_timer_sync(timer)
+
+#elif (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 91))
+
+#define timer_delete(timer)             del_timer(timer)
+
+#endif
+
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 3))
 
 #define PREPARE_TQUEUE(_tq, _routine, _data)			\
diff --git a/i386/src/wl/sys/wl_linux.c b/i386/src/wl/sys/wl_linux.c
index 77512a6..0397c3c 100644
--- a/i386/src/wl/sys/wl_linux.c
+++ b/i386/src/wl/sys/wl_linux.c
@@ -2415,7 +2415,7 @@ wl_del_timer(wl_info_t *wl, wl_timer_t *t)
 	ASSERT(t);
 	if (t->set) {
 		t->set = FALSE;
-		if (!del_timer(&t->timer)) {
+		if (!timer_delete(&t->timer)) {
 #ifdef BCMDBG
 			WL_INFORM(("wl%d: Failed to delete timer %s\n", wl->unit, t->name));
 #endif
